core/src/main/scala/org/ensime/core/javac/JavaCompletion.scala

```diff
@@ -0,0 +1,239 @@
+import com.sun.source.tree.Scope
+import scala.collection.mutable.HashSet;
+import scala.collection.mutable.ArrayBuffer;
+
+trait JavaCompletion { self: JavaCompiler =>
+
+  import CompletionUtil._
+
+  def completionsAt(info: SourceFileInfo, offset: Int, maxResultsArg: Int, caseSens: Boolean): CompletionInfoList = {
+    val s = contentsAsString(info, charset)
+
+    val preceding = s.slice(Math.max(0, offset - 100), offset)
+
+    System.err.println("PRECEDING: " + preceding)
+
+    val defaultPrefix = JavaIdentRegexp.findFirstMatchIn(preceding) match {
+      case Some(m) => m.group(1)
+      case _ => ""
+    }
+
+    System.err.println("PREFIX: " + defaultPrefix)
+
+    val constructing = ConstructingRegexp.findFirstMatchIn(preceding).isDefined
+
+    System.err.println("CONSTRUCTING: " + constructing)
+
+    val indexAfterTarget = offset - defaultPrefix.length - 1
+
+    val precedingChar = s(indexAfterTarget)
+
+    val isMemberAccess = precedingChar == '.'
+
+    (if (ImportSubtypeRegexp.findFirstMatchIn(preceding).isDefined) {
+      // Erase the trailing partial subtype (it breaks type resolution).
+      val patched = s.substring(0, indexAfterTarget) + " " + s.substring(indexAfterTarget + defaultPrefix.length + 1);
+      (pathToPoint(SourceFileInfo(info.file, Some(patched), None), indexAfterTarget - 1) map {
+        case (info: CompilationInfo, path: TreePath) => {
+          memberCandidates(info, path.getLeaf, defaultPrefix, true, caseSens)
+        }
+      })
+    } else if (ImportRegexp.findFirstMatchIn(preceding).isDefined) {
+      (pathToPoint(info, indexAfterTarget) flatMap {
+        case (info: CompilationInfo, path: TreePath) => {
+          getEnclosingMemberSelectTree(path).map { m =>
+            packageMemberCandidates(info, m, defaultPrefix, caseSens)
+          }
+        }
+      })
+    } else if (isMemberAccess) {
+      // TODO how to avoid allocating a new string? buffer of immutable string slices?
+      // Erase the trailing partial member (it breaks type resolution).
+      val patched = s.substring(0, indexAfterTarget) + ".wait()" + s.substring(indexAfterTarget + defaultPrefix.length + 1);
+      (pathToPoint(SourceFileInfo(info.file, Some(patched), None), indexAfterTarget) flatMap {
+        case (info: CompilationInfo, path: TreePath) => {
+          getEnclosingMemberSelectTree(path).map { m =>
+            memberCandidates(info, m.getExpression(), defaultPrefix, false, caseSens)
+          }
+        }
+      })
+    } else {
+      (scopeForPoint(info, indexAfterTarget) map {
+        case (info: CompilationInfo, s: Scope) => {
+          scopeMemberCandidates(info, s, defaultPrefix, caseSens)
+        }
+      })
+    }).getOrElse(CompletionInfoList(defaultPrefix, List()))
+  }
+
+  private def getEnclosingMemberSelectTree(path: TreePath): Option[MemberSelectTree] = {
+    var p = path
+    while (p != null) {
+      p.getLeaf match {
+        case m: MemberSelectTree => return Some(m)
```

Comment:
this is the first thing that comes up, but not sure it does a great job of motivating why **not** to use it https://tpolecat.github.io/2014/05/09/return.html

The problem I've seen in the wild is when somebody `return`s from a closure and it actually returns from the closure (whereas a natural reading would expect it to return the whole method) and continues through. Given that it's not always obvious when you're typing in a closure, or the method body, it results in all manner of weird stuff happening.